### PR TITLE
bugfix regex modes

### PR DESCRIPTION
**Pull request by thorinaboenke**
This PR solves issue #71 

- adds a regex.yml example file for all modes
- adds error handling to playbook parsing
- logs informative error when regex mode is sub and replace is missing
- slight refactoring of forge_variable() and sub_needs_replace(), reduce duplicated code, ensure first match is always named MATCH_0 (even if only one match exists), ensure command always returns the result of the re.<mode> commands, i.e. if there is no string to substitute, the original input is returned.
- adds unit test for RegExCommand and RegExExecutor
output of regex.yml playbook:

image

### CODE REVIEW

**Comment by thorinaboenke** [comment]
i leave this as draft until tests are completed

File: examples/regex.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    variable: $INPUT
+    cmd: "6667/tcp open  irc UnrealIRCd"
+
+  # MODE FINDALL
+  - type: regex
+    input: INPUT
+    cmd: "666"
+    mode: findall
+    output:
+        # {'MATCH_0': '6', 'MATCH_1': '6', 'MATCH_2': '6'}
```
Comment (whotwagner) [inline examples/regex.yml:21]: this shoud result to `{'MATCH_0': '666'}`

File: examples/regex.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    cmd: "Search Result: $SEARCH"
+
+    # MODE SUB
+  - type: regex
+    input: INPUT
+    replace: SUBSTITUTION
+    cmd: tcp
+    mode: sub
+    output:
+        # {'MATCH_0': '6667/hello open  irc UnrealIRCd'}
```
Comment (whotwagner) [inline examples/regex.yml:58]: I think "hello" should be "SUBSTITUTION" instead. right?

File: src/attackmate/schemas/regex.py [hunk 3]
```diff
@@ -2,21 +2,27 @@
 from pydantic import ValidationInfo, field_validator
 from .base import BaseCommand
 from pprint import pprint
+import logging
+
+logger = logging.getLogger('playbook')
```
Comment (whotwagner) [inline src/attackmate/schemas/regex.py:7]:
```suggestion
```

File: src/attackmate/schemas/regex.py [hunk 4]
```diff
@@ -2,21 +2,27 @@
...
     @field_validator('mode')
     @classmethod
     def sub_needs_replace(cls, v, info: ValidationInfo) -> str:
-        if 'replace' in info.data and info.data['replace']:
-            return v
-        else:
+
+        if v == 'sub' and not info.data.get('replace'):
             print(info.data)
             pprint(info)
```
Comment (whotwagner) [inline src/attackmate/schemas/regex.py:18]:
```suggestion
```

File: src/attackmate/schemas/regex.py [hunk 5]
```diff
@@ -2,21 +2,27 @@
...
-            raise ValueError('Regex: sub needs the replace-setting!')
+            logger.error('Error parsing playbook. command mode: sub must be preceded by replace setting!')
+            raise ValueError('Regex sub mode needs replace-setting!')
```
Comment (whotwagner) [inline src/attackmate/schemas/regex.py:20]:
```suggestion
            raise ValueError('Error parsing playbook. command mode: sub must be preceded by replace setting!)
```

**Review by whotwagner — CHANGES_REQUESTED** [review]

**Comment by thorinaboenke** [comment]
made the changes and adapted tests accordingly. ready for re-review

File: examples/regex.yml [hunk 6]
```diff
@@ -18,7 +18,7 @@ commands:
     cmd: "666"
     mode: findall
     output:
-        # {'MATCH_0': '6', 'MATCH_1': '6', 'MATCH_2': '6'}
+        # {'MATCH_0': '666'}
```
Comment (whotwagner) [inline examples/regex.yml:21]: The expected result is: `{'MATCH_0': '6', 'MATCH_1': '6', 'MATCH_2': '6'}`

**Review by whotwagner — CHANGES_REQUESTED** [review]

File: examples/regex.yml (hunk 6 above)
Comment (thorinaboenke) [reply examples/regex.yml:21]: @whotwagner , I actually changed this per your previous review, 
image
 i think 666 is the correct match? (if the findall would look for "6" only we would expect 3 results)

Comment (whotwagner) [reply examples/regex.yml:21]: Sorry, I made a mistake back then. The correct result would be 3 matches:

```
Python 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> var = "666"
>>> import re
>>> re.findall("6", var)
['6', '6', '6']
```

Comment (annaerdi) [reply examples/regex.yml:21]: But if it's `cmd: "666"` then it's doing a `re.findall("666", var)`, or?

Comment (whotwagner) [reply examples/regex.yml:21]: > But if it's `cmd: "666"` then it's doing a `re.findall("666", var)`, or?

true. sorry, my bad (again)

**Review by whotwagner — APPROVED** [review]

**Pull request merged**